Source/NameServer/AddressBook/AddressBook.cpp

```diff
@@ -0,0 +1,1367 @@
+        return(ERR_INVALID_ATTRIBUTE);
+    }
+
+    return SUCCESS;   // Validation Successful
+}
+
+/*==============================================================================
+ * CheckVector: Check that adding a Record to the vector won't cause a
+ *              reallocation as this would silently invalidate all of the maps.
+ *              If the vector is at capacity, increase it's size, clear the task
+ *              linkage & mappings and mark both as dirty.
+ *
+ *              As the Record vectors are pre-sized based on what should be in
+ *              the task plus one element, there should never be a reallocation.
+ *============================================================================*/
+unsigned AddressBook::CheckVector(TaskRecord_t *TRec, std::vector<Record_t> &Vect)
+{
+    if(Vect.size() == Vect.capacity())
+    {
+        Vect.reserve(static_cast<unsigned>(Vect.capacity()*1.25));
+        ClearLink(TRec);
+        ClearMap(TRec);
+        return ERR_INVALID_MAP;
+    }
+    return SUCCESS;
+}
+
+
+/*==============================================================================
+ * ClearMap:    Clear the Device to DeviceType/Attribute/(Message) Type links
+ *              and mark the linkage as dirty. Also clears the Name>Record map
+ *              and list of records with external connections.
+ *============================================================================*/
+unsigned AddressBook::ClearMap(TaskRecord_t *TRec)
+{
+    TRec->MapValid = false;     // Mark the map as dirty.
+    TRec->NameMap.clear();      // Clear the Name map
+    TRec->ExtCon.clear();       // Clear External Connections List.
+
+    // Clear DevType->DeviceRecord vectors
+    for(std::vector<DevTypePair>::iterator D=TRec->DevTypes.begin();
+           D!=TRec->DevTypes.end();D++)
+    {
+        D->second.clear();
+    }
+
+    // Clear Attribute->DeviceRecord vectors
+    for(std::vector<AttrTypePair>::iterator A=TRec->AttrTypes.begin();
+           A!=TRec->AttrTypes.end();A++)
+    {
+        A->second.clear();
+    }
+
+    return SUCCESS;
+}
+
+
+/*==============================================================================
+ * ClearLink:   Clear the Address>Record map and the Supervisor>Device links.
+ *              Also marks the linkage as dirty.
+ *============================================================================*/
+unsigned AddressBook::ClearLink(TaskRecord_t *TRec)
+{
+    TRec->LinkValid = false;    // Mark the linkage as dirty.
+    TRec->AddrMap.clear();      // Clear the Addres>DeviceRecord map.
+
+    //Clear the Device Record vectors for each supervisor.
+    for(SupMap_t::iterator S=TRec->SupMap.begin();S!=TRec->SupMap.end();S++)
+    {
+         S->second.clear();
+    }
+
+    return SUCCESS;
+}
+
+
+//==============================================================================
+//Integrity Checks
+unsigned AddressBook::IntegCheck(void)
```

Comment:
Not a fan of default arguments?

Reply:
not sure how it would work with a pointer to a vector if the vector doesn't exist?

Reply:
Like this, but you need to pass by pointer:

```
unsigned AddressBook::IntegCheck(std::vector<unsigned>* TaskIntegs=PNULL);

unsigned AddressBook::IntegCheck(std::vector<unsigned>* TaskIntegs)
{
    unsigned TInteg, Ret = 0;

    for(TaskMap_t::const_iterator T=TaskMap.begin();
            T!=TaskMap.end();T++)
    {
        TInteg = T->second->Integrity(false);
        if (TaskIntegs != PNULL) TaskIntegs->push_back(TInteg);
        if(TInteg) Ret++;
    }

    return Ret;
}
```

This means you don't need `unsigned AddressBook::IntegCheck(void)`.